google-cloud-tools-plugin/ultimate/src/com/google/cloud/tools/intellij/appengine/server/run/CloudSdkStartupPolicy.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.server.run;
+
+import com.google.cloud.tools.appengine.api.devserver.DefaultRunConfiguration;
+import com.google.cloud.tools.intellij.appengine.cloud.AppEngineExecutor;
+import com.google.cloud.tools.intellij.appengine.cloud.AppEngineStandardRunTask;
+import com.google.cloud.tools.intellij.appengine.server.instance.AppEngineServerModel;
+import com.google.cloud.tools.intellij.util.GctBundle;
+
+import com.intellij.execution.ExecutionException;
+import com.intellij.execution.process.OSProcessHandler;
+import com.intellij.execution.runners.ProgramRunner;
+import com.intellij.javaee.run.configuration.CommonModel;
+import com.intellij.javaee.run.localRun.EnvironmentHelper;
+import com.intellij.javaee.run.localRun.ExecutableObject;
+import com.intellij.javaee.run.localRun.ExecutableObjectStartupPolicy;
+import com.intellij.javaee.run.localRun.ScriptHelper;
+import com.intellij.javaee.run.localRun.ScriptsHelper;
+import com.intellij.packaging.artifacts.Artifact;
+
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
+
+/**
+ * Runs a Google App Engine Standard app locally with devappserver, through the tools lib.
+ */
+public class CloudSdkStartupPolicy implements ExecutableObjectStartupPolicy {
+
+  // The startup process handler is kept so the process can be explicitly terminated, since we're
+  // not delegating that to the framework.
+  OSProcessHandler startupProcessHandler;
+
+  @Nullable
+  @Override
+  public ScriptHelper createStartupScriptHelper(ProgramRunner programRunner) {
+    return new ScriptHelper() {
+      @Nullable
+      @Override
+      public ExecutableObject getDefaultScript(final CommonModel commonModel) {
+        return new ExecutableObject() {
+          @Override
+          public String getDisplayString() {
+            return GctBundle.getString("appengine.run.startupscript.name");
+          }
+
+          @Override
+          public OSProcessHandler createProcessHandler(String string, Map<String, String> map)
```

Comment:
I'm heading out. Thanks for taking all the feedback. Feel free to merge
when you've covered my last batch of comments.

Good work.

On Tue, Aug 23, 2016 at 6:20 PM, João Martins <notifications@github.com>
wrote:

> In google-cloud-tools-plugin/ultimate/src/com/google/cloud/
> tools/intellij/appengine/server/run/CloudSdkStartupPolicy.java
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/811#discussion_r75962495>
> :
>
> > +
> > +  @Nullable
> > +  @Override
> > +  public ScriptHelper createStartupScriptHelper(ProgramRunner programRunner) {
> > +    return new ScriptHelper() {
> > +      @Nullable
> > +      @Override
> > +      public ExecutableObject getDefaultScript(final CommonModel commonModel) {
> > +        return new ExecutableObject() {
> > +          @Override
> > +          public String getDisplayString() {
> > +            return GctBundle.getString("appengine.run.startupscript.name");
> > +          }
> > +
> > +          @Override
> > +          public OSProcessHandler createProcessHandler(String string, Map<String, String> map)
>
> Done. string -> workingDirectory, map -> envVariables.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/811/files/7ec4a74098d6c9952804ba728be7c059823bf0f5#r75962495>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HRRhjKQVekLo7uKIA6KHK3VMlhlxks5qi3IngaJpZM4JdGot>
> .
>